<div style="align: center; margin: 0; padding: 0; height: 250px;">
    <br>
    <img src="https://storage.googleapis.com/kaggle-datasets-images/5227/7876/3d18388d350d2791f4121a232acce097/dataset-cover.jpg" style="display:block; margin:auto; width:65%; height:100%;">
</div><br><br>

<div style="letter-spacing:normal; opacity:1.;">
<!--   https://xkcd.com/color/rgb/   -->
  <p style="text-align:center; background-color: lightsalmon; color: Jaguar; border-radius:10px; font-family:monospace; 
            line-height:1.4; font-size:32px; font-weight:bold; text-transform: uppercase; padding: 9px;">
            <strong>California Housing Prices Dataset</strong></p>  
            
  <p style="text-align:center; background-color:romance; color: Jaguar; border-radius:10px; font-family:monospace; 
            line-height:1.4; font-size:22px; font-weight:normal; text-transform: capitalize; padding: 5px;"
     >Machine Learning Engineer Module-1:<br>California Housing Prices Compute Weights<br><br></p>
    
  <div style="align: center;">
  <table style="text-align: center; background-color: romance; color: Jaguar; border-radius: 10px; font-family: monospace;
                  line-height:1.4; font-size: 21px; font-weight: normal; text-transform: capitalize; padding: 5px; 
                  margin: 0 auto;">
    <tr><td style="text-align: left; padding-left: 0px;"
            > Exploratory Data Analysis (EDA) <span style="font-size: 16px;"> </span></td></tr>
    <tr><td style="text-align: left; padding-left: 0px;"
            > Linear Regression with Functions <span style="font-size: 16px;">(Define Func: Implementation of Normal Equation for Compute Weights)</span></td></tr>
  </table>
  </div>

</div>

**Dataset More Info**

> https://www.kaggle.com/datasets/camnugent/california-housing-prices

**About Dataset**

**Context**

This is the dataset used in the second chapter of Aurélien Géron's recent book 'Hands-On Machine learning with Scikit-Learn and TensorFlow'. It serves as an excellent introduction to implementing machine learning algorithms because it requires rudimentary data cleaning, has an easily understandable list of variables and sits at an optimal size between being to toyish and too cumbersome.

The data contains information from the 1990 California census. So although it may not help you with predicting current housing prices like the Zillow Zestimate dataset, it does provide an accessible introductory dataset for teaching people about the basics of machine learning.

**Content**

The data pertains to the houses found in a given California district and some summary stats about them based on the 1990 census data. Be warned the data aren't cleaned so there are some preprocessing steps required! The columns are as follows, their names are pretty self explanitory:

longitude

latitude

housing_median_age

total_rooms

total_bedrooms

population

households

median_income

median_house_value

ocean_proximity

**Acknowledgements**

This data was initially featured in the following paper:

Pace, R. Kelley, and Ronald Barry. "Sparse spatial autoregressions." Statistics & Probability Letters 33.3 (1997): 291-297.

and I encountered it in 'Hands-On Machine learning with Scikit-Learn and TensorFlow' by Aurélien Géron.

Aurélien Géron wrote:

This dataset is a modified version of the California Housing dataset available from:

[Luís Torgo's page](http://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html) (University of Porto)

**Inspiration**

See my kernel on machine learning basics in R using this dataset, or venture over to the following link for a python based introductory tutorial:

https://github.com/ageron/handson-ml/tree/master/datasets/housing


[Download Data](https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv)

**TASK**

### Set up the environment

You need to install Python, NumPy, Pandas, Matplotlib and Seaborn. For that, you can the instructions from
[06-environment.md](https://github.com/alexeygrigorev/mlbookcamp-code/blob/master/course-zoomcamp/01-intro/06-environment.md).

## Submit the results

* Submit your results here: https://forms.gle/jneGM91mzDZ23i8HA
* You can submit your solution multiple times. In this case, only the last submission will be used 
* If your answer doesn't match options exactly, select the closest one


## Deadline

The deadline for submitting is 18 September 2023 (Monday), 23:00 CEST (Berlin time).

After that, the form will be closed.

Questions: https://github.com/DataTalksClub/machine-learning-zoomcamp/blob/master/cohorts/2023/01-intro/homework.md

<div style="letter-spacing:normal; opacity:1.;">
  <h1 style="text-align:center; background-color: lightsalmon; color: Jaguar; border-radius:10px; font-family:monospace; border-radius:20px;
            line-height:1.4; font-size:32px; font-weight:bold; text-transform: uppercase; padding: 9px;">
            <strong>1. Import Libraries & Ingest Data</strong></h1>   
</div>

> ⚠️ Not Recommended conda `base` env, work on `venv`

- https://docs.conda.io/projects/conda/en/4.6.0/_downloads/52a95608c49671267e40c689e0bc00ca/conda-cheatsheet.pdf

```
pip freeze > requirements.txt
conda list -e > requirements.txt

# new conda virtual environment
conda create --name "lin-reg" python=3.10 jupyter -y
conda activate "lin-reg"

# install all package dependencies
pip install -r requirements.txt
conda install -c conda-forge --file=requirements.txt      # mostly not work
conda install -c conda-forge pandas==2.0.2 -q -y

# if The environment is inconsistent, try below
conda update -n base -c defaults conda --force-reinstall
conda install anaconda --force-reinstall

```

**You must use the `--no-deps` option in the pip install command in order to avoid bundling dependencies into your conda-package.**

If you run pip install without the `--no-deps` option, pip will often install dependencies in your conda recipe and those dependencies will become part of your package. This wastes space in the package and `increases the risk of file overlap`, file clobbering, and broken packages.

There might be cases where you want to install a package directly from a local directory or a specific location, without relying on the package indexes. In such situations, you can use the `--no-index` option to tell pip not to look for the package in any indexes.

```
- command1 & command2  # runs simultaneously
- command1 ; command2  # runs sequentially
- command1 && command2 # runs sequentially, runs command2 only if command1 succeeds
- command1 || command2 # runs sequentially, runs command2 only if command1 fails
```

In [1]:
!python -V

Python 3.10.12


In [2]:
%%writefile requirements.txt 
# To get started with MLflow you'll need to install the appropriate Python package.

# for parquet file
pyarrow
fastparquet
# orjson is a fast, correct JSON library
orjson
numpy
pandas
matplotlib
seaborn
scipy
tqdm
joblib

# ML Model packages
scikit-learn
xgboost
optuna
hyperopt

# MLOPS experiment-tracking Packages
mlflow
wandb

# MLOPS Orchestration Packages
boto3
prefect
prefect-aws
prefect-email

# MLOPS Monitoring Packages
evidently
psycopg
psycopg[binary]

# Optionally
requests
# for py file parameter
click
# for code style
black

# Optionally
jupyter
ipykernel
ipywidgets
pipenv

Overwriting requirements.txt


In [3]:
# !pip install pyarrow fastparquet orjson pandas matplotlib seaborn tqdm requests scikit-learn xgboost optuna hyperopt
# !pip install mlflow prefect prefect-email boto3 prefect-aws joblib evidently
# !pip install psycopg psycopg[binary] pipenv jupyter ipykernel ipywidgets

In [4]:
import os, sys, platform, IPython.display

# pip install --no-deps --no-index --force-reinstall --no-cache-dir --user
# !{sys.executable} -m pip install -Uq -r requirements.txt --no-cache-dir --user
# !jupyter nbextension enable --py widgetsnbextension

# IPython.display.clear_output()
print("Platform:", platform.system())  # platform.platform()
print("Python  :", platform.python_version())  # sys.version
print("Actv Env:", os.getenv('CONDA_DEFAULT_ENV', 'Not Found Conda Env'))

Platform: Windows
Python  : 3.10.12
Actv Env: lin-reg


In [5]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

import statsmodels.api as sm
# import scipy
# import scipy.stats as stats
# from scipy.sparse import hstack
from scipy.stats import skew, norm, gamma, probplot

from sklearn.feature_extraction import DictVectorizer
from sklearn.preprocessing import OneHotEncoder
# from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler

from sklearn.linear_model import LinearRegression, TweedieRegressor
# from sklearn.linear_model import ElasticNet, Lasso, LassoCV, Ridge, RidgeCV
# from sklearn.svm import LinearSVR
from sklearn.ensemble import RandomForestRegressor
# from sklearn.ensemble import GradientBoostingRegressor, ExtraTreesRegressor

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
# from sklearn.model_selection import GridSearchCV

import xgboost as xgb

# Hyperparameter Optimization with Tree-structured Parzen Estimator (TPE) algorithm
import optuna
from hyperopt import fmin, tpe, hp, STATUS_OK, Trials
from hyperopt.pyll import scope

import os
import sys
import pickle
import zipfile
# import boto3
# import click
# import pathlib
# import argparse
# import requests
# import urllib.request
from glob import glob
# from tqdm import tqdm           # console-based
# from tqdm.notebook import tqdm  # jupyter-based
from tqdm.auto import tqdm        # automatically selects
# tqdm._instances.clear()
# Import joblib for model persistence
# from joblib import load, dump

# memory management performs garbage collection 
import gc
gc.collect()

0

In [6]:
# Filter out FutureWarnings
import warnings
# warnings.simplefilter(action='ignore', category=FutureWarning)
# warnings.filterwarnings(action="ignore", message="FutureWarning: is_categorical_dtype*")
warnings.filterwarnings(action="ignore", module="seaborn", category=FutureWarning, message='(is_categorical_dtype is deprecated)')
warnings.filterwarnings(action="ignore", module="seaborn", category=FutureWarning, message='(use_inf_as_na option is deprecated)')
warnings.filterwarnings(action="ignore", module="sklearn", category=FutureWarning, message='(is_sparse is deprecated)')
warnings.filterwarnings(action="ignore", module="xgboost", category=FutureWarning, message='(is_sparse is deprecated)')

### Question 1

What's the version of Pandas that you installed?

You can get the version information using the `__version__` field:

```python
pd.__version__
```

In [7]:
pd.__version__

'2.1.0'

In [8]:
# !pip install session_info
import session_info
session_info.show(html=False)

-----
hyperopt            0.2.7
matplotlib          3.7.2
numpy               1.25.2
optuna              3.3.0
pandas              2.1.0
scipy               1.11.2
seaborn             0.12.2
session_info        1.0.0
sklearn             1.2.2
statsmodels         0.14.0
tqdm                4.66.1
xgboost             1.7.6
-----
IPython             8.12.0
jupyter_client      7.4.9
jupyter_core        5.3.0
jupyterlab          3.6.3
notebook            6.5.4
-----
Python 3.10.12 | packaged by Anaconda, Inc. | (main, Jul  5 2023, 19:09:20) [MSC v.1916 64 bit (AMD64)]
Windows-10-10.0.22621-SP0
-----
Session information updated at 2023-11-01 15:44


In [9]:
import os
# Get the current working directory
# current_dir = os.getcwd()

# Create a new directory for storing data
# os.makedirs('./data', exist_ok=True)
# os.makedirs('./pycode', exist_ok=True)
# os.makedirs('./output', exist_ok=True)
# os.makedirs('./model', exist_ok=True)

<div style="letter-spacing:normal; opacity:1.;">
  <h1 style="text-align:center; background-color: lightsalmon; color: Jaguar; border-radius:10px; font-family:monospace; border-radius:20px;
            line-height:1.4; font-size:32px; font-weight:bold; text-transform: uppercase; padding: 9px;">
            <strong>2. Recognizing and Understanding Data</strong></h1>   
</div>

### Getting the data 

For this homework, we'll use the California Housing Prices dataset. Download it from 
[here](https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv).

You can do it with wget:

```bash
wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
```

Or just open it with your browser and click "Save as...".

Now read it with Pandas.

### Ingest Data [wget](https://linuxways.net/centos/linux-wget-command-with-examples/) or [curl](https://daniel.haxx.se/blog/2020/09/10/store-the-curl-output-over-there/)

In [10]:
# Download the data
!curl -Os --create-dirs --output-dir "./data" https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv

In [11]:
glob('data/*.*')

['data\\housing.csv']

In [12]:
# Read the data
csv_filename = 'housing.csv'
csv_filepath = f'data/{csv_filename}'

# Read the CSV file
df = pd.read_csv(csv_filepath, low_memory=False)

In [13]:
print("Shape of DataFrame:", df.shape)
df.head()

Shape of DataFrame: (20640, 10)


longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

### Question 2

How many columns are in the dataset?

- 10
- 6560
- 10989
- 20640

In [19]:
df.shape[1]

10

### Question 3

Which columns in the dataset have missing values?

- `total_rooms`
- `total_bedrooms`
- both of the above
- no empty columns in the dataset

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


In [17]:
df.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [21]:
df.isna().sum()[df.isna().sum()>0].index.tolist()

['total_bedrooms']

### Question 4

How many unique values does the `ocean_proximity` column have?

- 3
- 5
- 7
- 9

In [22]:
df.ocean_proximity.value_counts()

ocean_proximity
<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: count, dtype: int64

In [23]:
df.ocean_proximity.nunique()

5

In [16]:
df.describe(include='O').T

count unique        top  freq
ocean_proximity  20640      5  <1H OCEAN  9136

In [15]:
df.describe().T

count           mean            std         min  \
longitude           20640.0    -119.569704       2.003532   -124.3500   
latitude            20640.0      35.631861       2.135952     32.5400   
housing_median_age  20640.0      28.639486      12.585558      1.0000   
total_rooms         20640.0    2635.763081    2181.615252      2.0000   
total_bedrooms      20433.0     537.870553     421.385070      1.0000   
population          20640.0    1425.476744    1132.462122      3.0000   
households          20640.0     499.539680     382.329753      1.0000   
median_income       20640.0       3.870671       1.899822      0.4999   
median_house_value  20640.0  206855.816909  115395.615874  14999.0000   

                            25%          50%           75%          max  
longitude             -121.8000    -118.4900    -118.01000    -114.3100  
latitude                33.9300      34.2600      37.71000      41.9500  
housing_median_age      18.0000      29.0000      37.00000      52.0000  
total_rooms           1447.7500    2127.0000    3148.00000   39320.0000  
total_bedrooms         296.0000     435.0000     647.00000    6445.0000  
population             787.0000    1166.0000    1725.00000   35682.0000  
households             280.0000     409.0000     605.00000    6082.0000  
median_income            2.5634       3.5348       4.74325      15.0001  
median_house_value  119600.0000  179700.0000  264725.00000  500001.0000

<div style="letter-spacing:normal; opacity:1.;">
  <h1 style="text-align:center; background-color: lightsalmon; color: Jaguar; border-radius:10px; font-family:monospace; border-radius:20px;
            line-height:1.4; font-size:32px; font-weight:bold; text-transform: uppercase; padding: 9px;">
            <strong>3. Exloratory Data Analysis</strong></h1>   
</div>

### Question 5

What's the average value of the `median_house_value` for the houses located near the bay?

- 49433
- 124805
- 259212
- 380440

In [29]:
df[df.ocean_proximity == 'NEAR BAY']['median_house_value'].mean().astype('int')

259212

In [24]:
df.groupby('ocean_proximity')['median_house_value'].agg(['min', 'mean', 'max'])

min           mean       max
ocean_proximity                                   
<1H OCEAN         17500.0  240084.285464  500001.0
INLAND            14999.0  124805.392001  500001.0
ISLAND           287500.0  380440.000000  450000.0
NEAR BAY          22500.0  259212.311790  500001.0
NEAR OCEAN        22500.0  249433.977427  500001.0

### Question 6

1. Calculate the average of `total_bedrooms` column in the dataset.
2. Use the `fillna` method to fill the missing values in `total_bedrooms` with the mean value from the previous step.
3. Now, calculate the average of `total_bedrooms` again.
4. Has it changed?

Has it changed?

> Hint: take into account only 3 digits after the decimal point.

- Yes
- No

In [30]:
mean_bedrooms = df['total_bedrooms'].mean()
mean_bedrooms

537.8705525375618

In [31]:
df['total_bedrooms'].fillna(mean_bedrooms).mean()

537.8705525375617

**No, it's left the same.**

### Question 7

1. Select all the options located on islands.
2. Select only columns `housing_median_age`, `total_rooms`, `total_bedrooms`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[950, 1300, 800, 1000, 1300]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the value of the last element of `w`?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

- -1.4812
- 0.001
- 5.6992
- 23.1233

In [36]:
y = np.array([950, 1300, 800, 1000, 1300])
y.shape

(5,)

In [37]:
island_df = df[df.ocean_proximity == 'ISLAND']
island_df = island_df[['housing_median_age', 'total_rooms', 'total_bedrooms']]
island_df

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [38]:
X = island_df.values
X.shape

(5, 3)

In [41]:
X.T

array([[  27.,   52.,   52.,   52.,   29.],
       [1675., 2359., 2127.,  996.,  716.],
       [ 521.,  591.,  512.,  264.,  214.]])

In [44]:
XTX = X.T.dot(X)
XTX.shape

(3, 3)

In [45]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [46]:
w = XTX_inv.dot(X.T).dot(y)  # (XTX_inv @ X.T) @ y
w

array([23.12330961, -1.48124183,  5.69922946])

In [47]:
w[2]

5.699229455065586

> **Note**: we just implemented normal equation

$$w = (X^T X)^{-1} X^T y$$

## End of Project